firebase-installations/src/main/java/com/google/firebase/installations/Utils.java

```diff
@@ -0,0 +1,75 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.installations;
+
+import androidx.annotation.NonNull;
+import com.google.common.io.BaseEncoding;
+import java.nio.charset.Charset;
+import java.util.UUID;
+
+/** Util methods used for {@link FirebaseInstallations} */
+class Utils {
+
+  /**
+   * 1 Byte with the first 4 header-bits set to the identifying FID prefix 0111 (0x7). Use this
+   * constant to create FIDs or check the first byte of FIDs. This prefix is also used in legacy
+   * Instance-IDs
+   */
+  public static final byte FID_4BIT_PREFIX = Byte.parseByte("01110000", 2);
+
+  /**
+   * Byte mask to remove the 4 header-bits of a given Byte. Use this constant with Java's Binary AND
+   * Operator in order to remove the first 4 bits of a Byte and replacing it with the FID prefix.
+   */
+  public static final byte REMOVE_PREFIX_MASK = Byte.parseByte("00001111", 2);
+
+  /** Length of new-format FIDs. */
+  public static final int FID_LENGTH = 22;
+
+  /**
+   * Creates a random FID of valid format without checking if the FID is already in use by any
+   * Firebase Installation.
+   *
+   * <p>Note: Even though this method does not check with the FIS database if the returned FID is
+   * already in use, the probability of collision is extremely and negligibly small!
+   *
+   * @return random FID value
+   */
+  @NonNull
+  public static String createRandomFid() {
+    // A valid FID has exactly 22 base64 characters, which is 132 bits, or 16.5 bytes.
+    // We create 17 random bytes and ignore the last base64 character later.
+    byte[] bytes = UUID.randomUUID().toString().substring(0, 17).getBytes(Charset.defaultCharset());
```

Comment:
@diwu-arete  Please check if this is inline with [the comments on the doc](https://docs.google.com/a/google.com/document/d/1ib_AMf959UAPVyrzcnYMIFsxV2dMCzl5ZPG-8qHTH1M/edit?disco=AAAADFMTf9c)

Reply:
I think this is the right thing to do, and your encoding logic also looks good to me.

But definitely please ping Rayo to take a look at these key logic.

Reply:
@andirayo  Can you please review the random fid generation logic?

Thanks.

Reply:
This does NOT look good to me.
UUIDs are "only" 128bits, which is 16 bytes.
So something is going wrong here I think.
I don't think UUID.randomUUID().toString() can actually be 17 chars/bytes long...

I don't know what #randomUUID() does, or more precisely I don't know what #toString() does, but I fear that this may return the default UUID format containing dashes ("-").
Compare: https://en.wikipedia.org/wiki/Universally_unique_identifier#Format

=======================
Furthermore, I think the method #getBytes does not what you actually want to do here.
I'm not sure, but usually in Java calling #getBytes gets you the ASCii byte representation of a String, but not the byte-interpretation of a String. What I'm trying to say is: If your string is "a", then you want to get the Byte-value (integer 10), but I think #getBytes will get you (integer 97) which is the Ascii code of the character "a".
I'm happy to explain more in person.

=======================
Long story short, I think there are several things wrong in this line.
I would prefer to, if possible, avoid conversions to String. What is the point of converting anything to String here. Let's work with bits or bytes.

Reply:
Thanks for detailed explanation. As discussed through emails, I have implemented the FID creation as per Maksyms approach and now creates 2^128 FIDs. 

wdyt of this approach?